apps/argus/src/services/controller_service.rs

```diff
@@ -125,3 +161,595 @@ impl Service for ControllerService {
         Ok(())
     }
 }
+
+/// Determines if an on-chain price update is needed based on the latest Pyth price,
+/// the current on-chain price (if available), and the subscription's update criteria.
+#[tracing::instrument()]
+fn needs_update(
+    pyth_price: &Price,
+    chain_price_opt: Option<&Price>,
+    update_criteria: &UpdateCriteria,
+) -> bool {
+    // If there's no price currently on the chain for this feed, an update is always needed.
+    let chain_price = match chain_price_opt {
+        None => {
+            tracing::debug!("Update criteria met: No chain price available.");
+            return true;
+        }
+        Some(cp) => cp,
+    };
+
+    // 1. Heartbeat Check:
+    // Updates if `update_on_heartbeat` is enabled and the Pyth price is newer than or equal to
+    // the chain price plus `heartbeat_seconds`.
+    if update_criteria.update_on_heartbeat {
+        if pyth_price.publish_time
+            >= chain_price.publish_time + (update_criteria.heartbeat_seconds as i64)
+        {
+            tracing::debug!(
+                "Heartbeat criteria met: Pyth price is sufficiently newer or same age with met delta."
+            );
+            return true;
+        }
+    }
+
+    // 2. Deviation Check:
+    // If `update_on_deviation` is enabled, checks if the Pyth price has deviated from the chain price
+    // by more than `deviation_threshold_bps`.
+    // Example: If chain_price is 100 and deviation_threshold_bps is 50 (0.5%),
+    // then threshold_value = (100 * 50) / 10000 = 0.5
+    // This means a price difference of more than 0.5 would trigger an update
+    if update_criteria.update_on_deviation {
+        // Critical assumption: The `expo` fields of `pyth_price` and `chain_price` are identical,
+        // since we directly compare the `price` fields.
+        if chain_price.price == 0 {
+            if pyth_price.price != 0 {
+                tracing::debug!(
+                    "Deviation criteria met: Chain price is 0, Pyth price is non-zero."
+                );
+                return true;
```

Comment:
this is not in the on-chain logic, however the on-chain logic accepts any update if there's no update, maybe it is this one?